src/main/java/ch/naviqore/raptor/simple/Query.java

```diff
@@ -0,0 +1,214 @@
+        Set<Integer> markedStops = initialize();
+        markedStops.addAll(footpathRelaxer.relaxInitial(sourceStopIndices));
+        markedStops = removeSuboptimalLabelsForRound(0, markedStops);
+
+        // continue with further rounds as long as there are new marked stops
+        int round = 1;
+        while (!markedStops.isEmpty() && (round - 1) <= config.getMaximumTransferNumber()) {
+            // add label layer for new round
+            stopLabelsAndTimes.addNewRound();
+
+            // scan all routs and mark stops that have improved
+            Set<Integer> markedStopsNext = routeScanner.scan(round, markedStops);
+
+            // relax footpaths for all newly marked stops
+            markedStopsNext.addAll(footpathRelaxer.relax(round, markedStopsNext));
+
+            // prepare next round
+            markedStops = removeSuboptimalLabelsForRound(round, markedStopsNext);
+            round++;
+        }
+
+        return stopLabelsAndTimes.getBestLabelsPerRound();
+    }
+
+    /**
+     * Set up the best times per stop and best labels per round for a new query.
+     *
+     * @return the initially marked stops.
+     */
+    Set<Integer> initialize() {
+        log.info("Initializing global best times per stop and best labels per round");
+
+        // fill target stops
+        for (int i = 0; i < targetStops.length; i += 2) {
+            int index = (int) Math.ceil(i / 2.0);
+            targetStops[i] = targetStopIndices[index];
+            targetStops[i + 1] = walkingDurationsToTarget[index];
+        }
+
+        // set initial labels, best time and mark source stops
+        Set<Integer> markedStops = new HashSet<>();
+        for (int i = 0; i < sourceStopIndices.length; i++) {
+            int currentStopIdx = sourceStopIndices[i];
+            int targetTime = sourceTimes[i];
+
+            StopLabelsAndTimes.Label label = new StopLabelsAndTimes.Label(0, targetTime,
+                    StopLabelsAndTimes.LabelType.INITIAL, StopLabelsAndTimes.NO_INDEX, StopLabelsAndTimes.NO_INDEX,
+                    currentStopIdx, null);
+            stopLabelsAndTimes.setLabel(0, currentStopIdx, label);
+            stopLabelsAndTimes.setBestTime(currentStopIdx, targetTime);
+
+            markedStops.add(currentStopIdx);
+        }
+
+        return markedStops;
+    }
+
+    /**
+     * Nullify labels that are suboptimal for the current round. This method checks if the label time is worse than the
+     * optimal time mark and removes the mark for the next round and nullifies the label in this case.
+     *
+     * @param round       the round to remove suboptimal labels for.
+     * @param markedStops the marked stops to check for suboptimal labels.
+     */
+    Set<Integer> removeSuboptimalLabelsForRound(int round, Set<Integer> markedStops) {
+        int bestTime = getBestTimeForAllTargetStops();
+
+        if (bestTime == INFINITY || bestTime == -INFINITY) {
```

Comment:
@clukas1 do you need negative INFINITY in the simple raptor scenario ?

Reply:
Thanks for catching. Although the -INFINITY doesn't ever come true this is unnecessary. I removed it in commit 7c3137b and merged the branch.